readme.md
# ios-icons 
[![Build Status](https://travis-ci.com/randytarampi/ios-icons.svg?branch=master)](https://travis-ci.com/randytarampi/ios-icons)
[![Dependency Status](https://david-dm.org/randytarampi/ios-icons.svg)](https://david-dm.org/randytarampi/ios-icons)
[![devDependency Status](https://david-dm.org/randytarampi/ios-icons/dev-status.svg)](https://david-dm.org/randytarampi/ios-icons#info=devDependencies)
[![Semantically Released](https://img.shields.io/badge/versioning-semantically%20released-brightgreen.svg)](https://github.com/boennemann/semantic-release) 

> Get iOS icon file names and dimensions (width)

The default icon file names and required sizes for iOS are listed in a [JSON file](icons.json). This information is useful, for example, when you want to generate icons with the required sizes or to create a [`config.xml`](http://docs.phonegap.com/en/3.5.0/config_ref_images.md.html) file for a PhoneGap/Cordova project or if you just need to create the icons for your iOS project from one source icon.
This information also exists for [android icons](https://github.com/randytarampi/android-icons).

## Install

```sh
$ npm install --save @randy.tarampi/ios-icons
```


## Usage

```js
var icons = require('@randy.tarampi/ios-icons');

icons();
//=> [{name: 'icon-60@3x.png', width: 180}, {name: 'icon-60.png', width: 60, ...}]

icons({size: '72'})
//=> {name: 'icon-72.png', width: 72}

icons({size: 'small'})
//=> {name: 'icon-small.png', width: 29}
```


## API

### icons()

Returns an array of icons, each icon being represented by an object with `name` and `width` properties.

> Notice that icons are always squares, so no `height` property is provided.

### icons(options)
#### options

Only option for now is `size`, which can be either a `Number` or `String` value. If it is a `Number`, it represents the width in pixels. If it is a `String`, you can use `"@2x"` or `"@3x"` notation to refer to a certain size or the complete file name, e.g. `icon-small.png`.

Returns icon object for that size or `null`.

For example:

```js
icons({size: "60@3x"});
// ==> {name: "icon-60@3x.png", width: 180}
```


## CLI
> ios-icons logs to stdout in comma-separated values format (csv) by default so you can easy pipe to other commands in UNIX systems.

```sh
$ npm install --global @randy.tarampi/ios-icons
```

```sh
$ ios-icons --help

Examples:
    $ ios-icons --format json --size 60
    { name: 'icon-60.png', width: 60 }

    $ ios-icons --size 60@3x
    icon-60@3x.png,180

    $ ios-icons
    icon-60@3x.png,180
    icon-60.png,60
    icon-60@2x.png,120
    icon-76.png,76
    icon-76@2x.png,152
    icon-40.png,40
    icon-40@2x.png,80
    icon.png,57
    icon@2x.png,114
    icon-72.png,72
    icon-72@2x.png,144
    icon-small.png,29
    icon-small@2x.png,58
    icon-50.png,50
    icon-50@2x.png,100

    $ ios-icons --format json
    [{"name":"icon-60@3x.png","width":180},{"name":"icon-60.png","width":60},{"name":"icon-60@2x.png","width":120},{"name":"icon-76.png","width":76},{"name":"icon-76@2x.png","width":152},{"name":"icon-40.png","width":40},{"name":"icon-40@2x.png","width":80},{"name":"icon.png","width":57},{"name":"icon@2x.png","width":114},{"name":"icon-72.png","width":72},{"name":"icon-72@2x.png","width":144},{"name":"icon-small.png","width":29},{"name":"icon-small@2x.png","width":58},{"name":"icon-50.png","width":50},{"name":"icon-50@2x.png","width":100}]
```

## Semantic Releases
Since version 2.0.0, this module is being [semantically released](https://github.com/boennemann/semantic-release). You can safely use `"^2.0.0"` in your `package.json`.

## License
This was originally based on [animals](https://github.com/boennemann/animals) by [Stephan Bönnemann](http://boennemann.me/).
MIT © [David Pfahler](http://excellenteasy.com)


CHANGELOG.md
## [4.2.2](https://github.com/randytarampi/ios-icons/compare/v4.2.1...v4.2.2) (2020-04-30)


### Bug Fixes

* upgrade @semantic-release/changelog from 5.0.0 to 5.0.1 ([9e1b6f4](https://github.com/randytarampi/ios-icons/commit/9e1b6f492619f4dea7063f3562f8b304d0dc5147))
* upgrade @semantic-release/changelog from 5.0.0 to 5.0.1 ([1f377fe](https://github.com/randytarampi/ios-icons/commit/1f377fede421f63736ba27f7c11aab62d711dd63))
* upgrade standard from 14.3.1 to 14.3.3 ([b6f96f4](https://github.com/randytarampi/ios-icons/commit/b6f96f4e59ce10437161463f3d0e06407f700972))
* upgrade standard from 14.3.1 to 14.3.3 ([f736012](https://github.com/randytarampi/ios-icons/commit/f7360127b74018615d0b29a88737161247dc73de))
* upgrade yargs from 15.3.0 to 15.3.1 ([c81d325](https://github.com/randytarampi/ios-icons/commit/c81d3259cd7a16f16eda6aeeb615ba33b4a35839))
* upgrade yargs from 15.3.0 to 15.3.1 ([4790d34](https://github.com/randytarampi/ios-icons/commit/4790d34b9c4eef6b6e5ce5a1a1aed4aff4d20df9))

## [4.2.1](https://github.com/randytarampi/ios-icons/compare/v4.2.0...v4.2.1) (2020-04-25)


### Bug Fixes

* package.json & package-lock.json to reduce vulnerabilities ([7600cf4](https://github.com/randytarampi/ios-icons/commit/7600cf46c7ff7551df432da0cb1cb8387657cff9))

# [4.2.0](https://github.com/randytarampi/ios-icons/compare/v4.1.0...v4.2.0) (2020-01-06)


### Features

* **package:** Blindly upgrade our dependencies for the new year and support node@10, node@12 and beyond. ([e81fbb8](https://github.com/randytarampi/ios-icons/commit/e81fbb8de2422757bd0938447f6690dad62e9b6f))

# [4.1.0](https://github.com/randytarampi/ios-icons/compare/v4.0.0...v4.1.0) (2019-03-23)


### Features

* Also generate `icon-40@3x`. ([1c09dc2](https://github.com/randytarampi/ios-icons/commit/1c09dc2))

# [4.0.0](https://github.com/randytarampi/ios-icons/compare/v3.4.0...v4.0.0) (2019-03-10)


### Bug Fixes

* **package:** Since the package updates we no longer support node@4. ([f2227bc](https://github.com/randytarampi/ios-icons/commit/f2227bc))
* **travis:** Don't bother caching dependencies. ([f77b9e3](https://github.com/randytarampi/ios-icons/commit/f77b9e3))


### BREAKING CHANGES

* **package:** Drop support for node 4 and earlier, since `yargs^13.2.2` requires ES6 support.

# [3.4.0](https://github.com/randytarampi/ios-icons/compare/v3.3.3...v3.4.0) (2019-03-10)


### Features

* Add the App Store icon. ([ac315e1](https://github.com/randytarampi/ios-icons/commit/ac315e1))
* **travis:** `ios-icons` -> `[@randy](https://github.com/randy).tarampi/ios-icons` ([e7cf94e](https://github.com/randytarampi/ios-icons/commit/e7cf94e))


test.js
'use strict'
var test = require('tape')
var iosIcons = require('./')
var exec = require('child_process').exec

test('returns all icons in array', function (t) {
  t.plan(2)
  var icons = iosIcons()
  t.ok(Array.isArray(icons), 'returned an array')
  t.equal(icons.length, icons.length, icons.length + ' icons returned')
})

test('returns icon for size 60 as Number', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: 60 })
  t.ok(icon.name === 'icon-60.png', 'icon name correct')
  t.ok(icon.width === 60, 'icon width correct')
})

test('returns icon for size 60 as String', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: '60' })
  t.ok(icon.name === 'icon-60.png', 'icon name correct')
  t.ok(icon.width === 60, 'icon width correct')
})

test('returns icon for size 120', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: 120 })
  t.ok(icon.name === 'icon-60@2x.png', 'icon name correct')
  t.ok(icon.width === 120, 'icon width correct')
})

test('returns icon for size 83.5', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: 167 })
  t.ok(icon.name === 'icon-83.5@2x.png', 'icon name correct')
  t.ok(icon.width === 167, 'icon width correct')
})

test('returns icon for size 60@2x', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: '60@2x' })
  t.ok(icon.name === 'icon-60@2x.png', 'icon name correct')
  t.ok(icon.width === 120, 'icon width correct')
})

test('returns icon for size 60@3x', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: '60@3x' })
  t.ok(icon.name === 'icon-60@3x.png', 'icon name correct')
  t.ok(icon.width === 180, 'icon width correct')
})

test('returns icon for empty string with width 57', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: '' })
  t.ok(icon.name === 'icon.png', 'icon name correct')
  t.ok(icon.width === 57, 'icon width correct')
})

test('returns icon for size 57 as Number', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: 57 })
  t.ok(icon.name === 'icon.png', 'icon name correct')
  t.ok(icon.width === 57, 'icon width correct')
})

test('returns icon for size 57 as String', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: '57' })
  t.ok(icon.name === 'icon.png', 'icon name correct')
  t.ok(icon.width === 57, 'icon width correct')
})

test('returns icon for @2x', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: '@2x' })
  t.ok(icon.name === 'icon@2x.png', 'icon name correct')
  t.ok(icon.width === 114, 'icon width correct')
})

test('returns icon for size 114 as Number', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: 114 })
  t.ok(icon.name === 'icon@2x.png', 'icon name correct')
  t.ok(icon.width === 114, 'icon width correct')
})

test('returns icon for size 114 as String', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: '114' })
  t.ok(icon.name === 'icon@2x.png', 'icon name correct')
  t.ok(icon.width === 114, 'icon width correct')
})

test('returns null for size 123', function (t) {
  t.plan(1)
  var icon = iosIcons({ size: 123 })
  t.ok(icon === null)
})

test('returns icon for size 29 as small', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: 29 })
  t.ok(icon.name === 'icon-small.png', 'icon name correct')
  t.ok(icon.width === 29, 'icon width correct')
})

test('returns icon for size 58 as small@2x', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: 58 })
  t.ok(icon.name === 'icon-small@2x.png', 'icon name correct')
  t.ok(icon.width === 58, 'icon width correct')
})

test('returns icon for file name icon-small@2x.png', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: 'icon-small@2x.png' })
  t.ok(icon.name === 'icon-small@2x.png', 'icon name correct')
  t.ok(icon.width === 58, 'icon width correct')
})

test('returns icon for file name icon-small@3x.png', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: 'icon-small@3x.png' })
  t.ok(icon.name === 'icon-small@3x.png', 'icon name correct')
  t.ok(icon.width === 87, 'icon width correct')
})

test('returns icon for file name icon.png', function (t) {
  t.plan(2)
  var icon = iosIcons({ size: 'icon.png' })
  t.ok(icon.name === 'icon.png', 'icon name correct')
  t.ok(icon.width === 57, 'icon width correct')
})

test('cli returns all images as csv', function (t) {
  t.plan(1)
  var expected = [
    'icon-60@3x.png,180',
    'icon-60.png,60',
    'icon-60@2x.png,120',
    'icon-76.png,76',
    'icon-76@2x.png,152',
    'icon-83.5@2x.png,167',
    'icon-40.png,40',
    'icon-40@2x.png,80',
    'icon-40@3x.png,120',
    'icon.png,57',
    'icon@2x.png,114',
    'icon-72.png,72',
    'icon-72@2x.png,144',
    'icon-small.png,29',
    'icon-small@2x.png,58',
    'icon-small@3x.png,87',
    'icon-50.png,50',
    'icon-50@2x.png,100',
    'icon-1024.png,1024\n'
  ].join('\n')
  exec('./bin/ios-icons.js', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})

test('cli returns image with specified size as csv', function (t) {
  t.plan(1)
  var expected = 'icon-60@3x.png,180\n'
  exec('./bin/ios-icons.js --size 180', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})

test('cli returns image with specified size as csv with abbreviated flags', function (t) {
  t.plan(3)
  var expected = 'icon-60@3x.png,180\n'
  exec('./bin/ios-icons.js --siz 180', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
  exec('./bin/ios-icons.js --si 180', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
  exec('./bin/ios-icons.js --s 180', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})

test('cli returns correct image for size "60@3x" as json', function (t) {
  t.plan(1)
  var expected = '{"name":"icon-60@3x.png","width":180}\n'
  exec('./bin/ios-icons.js --size 60@3x --format json', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})

test('cli returns correct image for size "60@3x" as json w/ abbreviated flags', function (t) {
  t.plan(2)
  var expected = '{"name":"icon-60@3x.png","width":180}\n'
  exec('./bin/ios-icons.js --size 60@3x --forma json', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
  exec('./bin/ios-icons.js --size 60@3x --f json', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})


index.js
'use strict'
var icons = require('./icons.json')
var idRegEx = /^icon-?(.*)\.png$/
var widths = icons.map(function (icon) {
  return icon.width
})
var ids = icons.map(function (icon) {
  return icon.name.match(idRegEx)[1]
})

function getIconForSize (size) {
  var width = getWidthForSize(size)

  if (!width) {
    return null
  }

  return icons[widths.indexOf(width)] || null
}

function getWidthForSize (size) {
  if (typeof size === 'number') {
    return size
  }
  var width = Number(size)
  if (!isNaN(width) && size !== '') {
    return width
  }
  width = widths[ids.indexOf(size)]
  if (width) {
    return width
  }
  return widths[ids.indexOf(size.match(idRegEx)[1])]
}

module.exports = function (options) {
  options = options || {}
  var size = options.size
  if (!size && size !== '') {
    return icons
  }
  return getIconForSize(size)
}

module.exports.icons = icons
module.exports.getIconForSize = getIconForSize
module.exports.getWidthForSize = getWidthForSize


cli.js
#!/usr/bin/env node
'use strict'
var abbrev = require('abbrev')
var argv = require('minimist')(process.argv.slice(2), abbrev('help', 'version', 'size', 'format'))
var pkg = require('./package.json')
var icons = require('./')

function help () {
  console.log([
    pkg.description,
    '',
    'Use `--format json` to set output to JSON.',
    'Get specifc icon by size or name by using `--size`.',
    '',
    'Examples:',
    '  $ ios-icons --size 80',
    '  icon-40@2x.png,80',
    '',
    '  $ ios-icons --size 80 --format json',
    '  {"name":"icon-40@2x.png","width":80}',
    '',
    '  $ ios-icons --size small',
    '  icon-small.png,29'
  ].join('\n'))
}

function formatLog (icons, argv) {
  var format = (argv.format || 'csv').toLowerCase()
  if (format === 'json') {
    return JSON.stringify(icons)
  }
  if (!Array.isArray(icons)) {
    icons = [icons]
  }
  return icons.map(function (icon) {
    return icon.name + ',' + icon.width
  }).join('\n')
}

function cli () {
  if (argv.help) return help()

  if (argv.version) return console.log(pkg.version)

  var options = {
    size: argv.size || argv.s
  }

  var output = icons(options)
  if (output) console.log(formatLog(output, argv))
}

cli()
